### ISSUE TITLE
[SECURITY] Releases are built/executed/released in the context of insecure/untrusted code

### ISSUE THREAD

**Issue by JLLeitschuh** [paypal/NNAnalytics#198]
[CWE-829: Inclusion of Functionality from Untrusted Control Sphere](https://cwe.mitre.org/data/definitions/829.html)
[CWE-494: Download of Code Without Integrity Check](https://cwe.mitre.org/data/definitions/494.html)

The build files indicate that this project is resolving dependencies over HTTP instead of HTTPS. Any of these artifacts could have been MITM to maliciously compromise them and infect the build artifacts that were produced. Additionally, if any of these JARs or other dependencies were compromised, any developers using these could continue to be infected past updating to fix this.

This vulnerability has a CVSS v3.0 Base Score of 8.1/10
https://nvd.nist.gov/vuln-metrics/cvss/v3-calculator?vector=AV:N/AC:H/PR:N/UI:N/S:U/C:H/I:H/A:H

### This isn't just theoretical
POC code has existed since 2014 to maliciously compromise a JAR file inflight.
See:
* https://max.computer/blog/how-to-take-over-the-computer-of-any-java-or-clojure-or-scala-developer/
* https://github.com/mveytsman/dilettante

### MITM Attacks Increasingly Common
See:
* https://serverfault.com/a/153065
* https://security.stackexchange.com/a/12050
* [Comcast continues to inject its own code into websites you visit](https://thenextweb.com/insights/2017/12/11/comcast-continues-to-inject-its-own-code-into-websites-you-visit/#) (over HTTP)

### These Vulnerabilities are Considered "Out Of Scope" by the PayPal Security Team

I originally reported this vulnerability privately to the PayPal bug bounty program but it was determined to be "Out Of Scope" by the HackerOne team that pre-filters these reports.

If this is considered a security vulnerability, I'd advise the PayPal team to communicate this to the HackerOne team.

Here is the link to my original report:
https://hackerone.com/reports/504119

#### Insecure Download
https://github.com/paypal/NNAnalytics/blob/4701020d582e002fe6982483adf86f5d99298d65/build.gradle#L136-L138

https://github.com/paypal/gimel/blob/33192730edd6741ef459fba98b84e20d4f3ff980/pom.xml#L42-L58

https://github.com/paypal/spring-social-security-paypal-openidconnect-quickstart/blob/6a57453c67e404652f966f8414943008ec2da74c/pom.xml#L203-L224

https://github.com/paypal/aurora/blob/a2021e6829fb6024eac45f4afe54e2fbc70b847c/grails-app/conf/BuildConfig.groovy#L45

**WARNING!** If any of these builds are using a shared or re-used `~/.gradle` or `~/.m2` cache between builds and any of these downloads were maliciously compromised, the compromised jar may remain inside of cache directory and continue to be used in the future.

#### Insecure Upload
Passwords in this upload are being sent in plaintext and should be considered compromised!

https://github.com/paypal/digraph-parser/blob/3a947dd99e521c23da657bd56cb85f46bca56fa6/pom.xml#L21-L30

https://github.com/paypal/Gibberish-Detector-Java/blob/034bbbcce7bf5d4099e02a64362b31bddbbb29de/pom.xml#L40-L43

https://github.com/paypal/SeLion/blob/248163ce623b3d8a6c0444c677a2cedce2aec823/pom.xml#L70-L73

### Fix and Public Disclosure

At a minimum, all of these code locations where artifacts are downloaded from an untrusted source needs to be fixed. Previous releases should be rebuilt with the fix applied. The checksum of the released artifacts and artifacts built in a trusted environment should be made. If the checksums match, you can be certain that they weren't compromised.

If the checksums don't match, indicating a compromised artifact, CVE numbers need to be issued for the potentially malicious artifacts.

The ability to check if checksums match assume that these projects have [reproducible builds](https://en.wikipedia.org/wiki/Reproducible_builds).

**Comment by pjeli** [comment]
Hi @JLLeitschuh. Very nice finds here. I just want to make sure I understand this fully as it seems there are some items here that are not related to NameNode Analytics.

As far as I can understand from your report, the only highlighted item for my project appears to be here: NNAnalytics/build.gradle -- Lines 136 to 138 in 4701020. Is this correct? If so this is not a major concern as the HortonWorks repo is maintained by Hortonworks (Cloudera now) and used by all HortonWorks customers for fetching JARs etc. If there was an incident here it would be very far spread. That being said it is also trivial for me to update the build.gradle and make it fetch over HTTPS.

Please let me know if there are any other outstanding security issues. As far as I can see there is nothing else that pertains to NNA in your report but perhaps I am wrong.

**Comment by pjeli** [comment]
All release builds of NNA come with build information in the version.properties file, located under `/usr/local/nn-analytics/version.properties`.  However they do not come with a checksum. I will use this case to add checksums to the builds as well.

**Comment by pjeli** [comment]
I found your comments here regarding the same issue: https://github.com/paypal/Gibberish-Detector-Java/issues/2

I will change the password on the build deployments in Travis CI to make sure everything is OK and double check the checksums of the released builds.

**Comment by pjeli** [comment]
Committed change to fetch HortonWorks JARs over HTTPS.
Checking over previous releases now.

**Comment by JLLeitschuh** [comment]
@pjeli There's another issue with the HortonWorks repository where that repo is mirroring almost every single artifact over HTTP instead of HTTPS. That is an infrastructure vulnerability for all of Hortonworks.

https://repo.hortonworks.com/index.html#view-repositories

There are several other security vulnerabilities I'm finding in HortonWorks infrastructure that I'd rather discuss in a private form. Does HortonWorks have a security email that I can reach out to responsibly disclose these?

Please reach out to me at my email address (listed on my GitHub profile) to discuss.

**Comment by pjeli** [comment]
I have emailed you directly. Thanks.

**Comment by pjeli** [comment]
Upon further examination I don't believe I am even using the HortonWorks repo. I am actually pulling in Apache dependencies over Maven Central, not HortonWorks JARs.

I think the URL is left over artifacts from a build.gradle I had re-used from a much older project. I will make another change shortly to remove the URL / repo from the build.

**Comment by pjeli** [comment]
I am attaching a script I ran, utilizing `pkgdiff` utility, along with the outputs, showing that the release artifacts and the builds are clean.

Within the pkgdiff_reports directory you will find an HTML file for all `rpm` builds, showing a difference of two files: `version.properties` and `hadoop-nna-util-1.0-SNAPSHOT.jar`. The reasons for these differences are due to the differences in build environments and are harmless to the build. I also show that if you unpack the RPM and compare the `hadoop-nna-util-1.0-SNAPSHOT.jar` JARs directly there is no difference.

[compareRpms.sh.zip](https://github.com/paypal/NNAnalytics/files/2928723/compareRpms.sh.zip)
[pkgdiff_reports.zip](https://github.com/paypal/NNAnalytics/files/2928724/pkgdiff_reports.zip)

Closing this for now.
